tests/Emulation/Client/PrivateKey.hs

```diff
@@ -7,57 +7,74 @@ module Emulation.Client.PrivateKey
     toECDSAKey,
     toByteString,
     toCOSEAlgorithmIdentifier,
+    toRSAKey,
   )
 where
 
 import qualified Crypto.Hash as Hash
 import qualified Crypto.PubKey.ECC.ECDSA as ECDSA
 import qualified Crypto.PubKey.Ed25519 as Ed25519
+import qualified Crypto.PubKey.RSA as RSA
+import qualified Crypto.PubKey.RSA.PKCS15 as RSA
 import Crypto.Random (MonadRandom)
-import Crypto.WebAuthn.PublicKey
-  ( COSEAlgorithmIdentifier
-      ( COSEAlgorithmIdentifierES256,
-        COSEAlgorithmIdentifierES384,
-        COSEAlgorithmIdentifierES512,
-        COSEAlgorithmIdentifierEdDSA
-      ),
-  )
 import qualified Crypto.WebAuthn.PublicKey as PublicKey
 import qualified Data.ASN1.BinaryEncoding as ASN1
 import qualified Data.ASN1.Encoding as ASN1
 import qualified Data.ASN1.Prim as ASN1
 import qualified Data.ByteArray as BA
 import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
 
 data PrivateKey
   = ES256PrivateKey ECDSA.PrivateKey
   | ES384PrivateKey ECDSA.PrivateKey
   | ES512PrivateKey ECDSA.PrivateKey
   | Ed25519PrivateKey Ed25519.SecretKey
+  | RS1PrivateKey RSA.PrivateKey
+  | RS256PrivateKey RSA.PrivateKey
+  | RS384PrivateKey RSA.PrivateKey
+  | RS512PrivateKey RSA.PrivateKey
   deriving (Eq, Show)
 
 data Signature
   = ESSignature ECDSA.Signature
   | Ed25519Signature Ed25519.Signature
+  | RSSignature BS.ByteString
 
 toECDSAKey :: MonadFail f => PublicKey.COSEAlgorithmIdentifier -> ECDSA.PrivateKey -> f PrivateKey
 toECDSAKey PublicKey.COSEAlgorithmIdentifierES256 = pure . ES256PrivateKey
 toECDSAKey PublicKey.COSEAlgorithmIdentifierES384 = pure . ES384PrivateKey
 toECDSAKey PublicKey.COSEAlgorithmIdentifierES512 = pure . ES512PrivateKey
-toECDSAKey _ = const $ fail "Not a ECDSA key identifier"
+toECDSAKey _ = const $ fail "Not an ECDSA key identifier"
```

Comment:
It's a bit inconsistent that in some cases `fail` is used, but in others `error` is used, seemingly for the same class of errors, that is, errors that can only occur when the module is used wrongly or has a bug. I'd think `error` is more appropriate in these cases

Reply:
Hmm, it seems a bit messy, but alright. I should point out that this shouldn't cause any DOS, because this would just error out a thread for a single request. And servers will handle errors somewhere along the stack (or ignore them).